diff -Naur a/include/configs/am335x_evm.h b/include/configs/am335x_evm.h
--- a/include/configs/am335x_evm.h	2013-02-18 11:02:32.000000000 +0100
+++ b/include/configs/am335x_evm.h	2013-02-18 11:11:59.000000000 +0100
@@ -170,7 +170,7 @@
  /* Physical Memory Map */
 #define CONFIG_NR_DRAM_BANKS		1		/* 1 bank of DRAM */
 #define PHYS_DRAM_1			0x80000000	/* DRAM Bank #1 */
-#define PHYS_DRAM_1_SIZE		0x10000000	/* 256 MiB */
+#define PHYS_DRAM_1_SIZE		0x08000000	/* 128 MiB */
 
 #define CONFIG_SYS_SDRAM_BASE		PHYS_DRAM_1
 #define CONFIG_SYS_INIT_RAM_ADDR	SRAM0_START
