--- a/batt.c	2012-12-14 09:36:54.647605625 +0100
+++ b/batt.c	2012-12-14 09:38:17.615193286 +0100
@@ -95,7 +95,7 @@ typedef struct {
 typedef struct {
     char *command;
     sem_t *lock;
-} alarm;
+} alarm_batt;
 
 static void destructor(Plugin *p);
 static void update_display(lx_battery *lx_b, gboolean repaint);
@@ -103,7 +103,7 @@ static void update_display(lx_battery *l
 /* alarmProcess takes the address of a dynamically allocated alarm struct (which
    it must free). It ensures that alarm commands do not run concurrently. */
 static void * alarmProcess(void *arg) {
-    alarm *a = (alarm *) arg;
+    alarm_batt *a = (alarm_batt *) arg;
 
     sem_wait(a->lock);
     system(a->command);
@@ -157,7 +157,7 @@ void update_display(lx_battery *lx_b, gb
 	/* Run the alarm command if it isn't already running */
 	if (alarmCanRun) {
 	    
-	    alarm *a = (alarm *) malloc(sizeof(alarm));
+	    alarm_batt *a = (alarm_batt *) malloc(sizeof(alarm_batt));
 	    a->command = lx_b->alarmCommand;
 	    a->lock = &(lx_b->alarmProcessLock);
 	    
