--- linux-orig/arch/arm/Kconfig	2012-11-09 16:18:21.000000000 +0100
+++ linux-gisbc/arch/arm/Kconfig	2012-12-12 15:18:35.514651211 +0100
@@ -1710,7 +1710,7 @@ config FORCE_MAX_ZONEORDER
 	int "Maximum zone order" if ARCH_SHMOBILE
 	range 11 64 if ARCH_SHMOBILE
 	default "9" if SA1111
-	default "11"
+	default "12"
 	help
 	  The kernel memory allocator divides physically contiguous memory
 	  blocks into "zones", where each zone is a power of two number of
