diff -rupN orig/arch/arm/mach-omap2/id.c mod/arch/arm/mach-omap2/id.c
--- orig/arch/arm/mach-omap2/id.c	2014-04-15 19:03:00.000000000 +0200
+++ mod/arch/arm/mach-omap2/id.c	2014-05-04 13:42:02.000000000 +0200
@@ -409,8 +409,22 @@ void __init omap3xxx_check_revision(void
 		}
 		break;
 	case 0xb944:
-		omap_revision = AM335X_REV_ES1_0;
-		cpu_rev = "1.0";
+		switch (rev) {
+		case 0:
+			omap_revision = AM335X_REV_ES1_0;
+			cpu_rev = "1.0";
+			break;
+		case 1:
+			omap_revision = AM335X_REV_ES2_0;
+			cpu_rev = "2.0";
+			break;
+		case 2:
+		/* FALLTHROUGH */
+		default:
+			omap_revision = AM335X_REV_ES2_1;
+			cpu_rev = "2.1";
+			break;
+		}
 		break;
 	case 0xb8f2:
 		switch (rev) {
diff -rupN orig/arch/arm/mach-omap2/opp3xxx_data.c mod/arch/arm/mach-omap2/opp3xxx_data.c
--- orig/arch/arm/mach-omap2/opp3xxx_data.c	2014-04-15 19:06:00.000000000 +0200
+++ mod/arch/arm/mach-omap2/opp3xxx_data.c	2014-05-03 01:51:10.000000000 +0200
@@ -18,6 +18,7 @@
  * GNU General Public License for more details.
  */
 #include <linux/module.h>
+#include <linux/io.h>
 
 #include <plat/cpu.h>
 
@@ -167,7 +168,7 @@ static struct omap_opp_def __initdata om
 #define AM33XX_VDD_MPU_OPP120_UV	1200000
 #define AM33XX_VDD_MPU_OPPTURBO_UV	1260000
 
-static struct omap_opp_def __initdata am33xx_opp_def_list[] = {
+static struct omap_opp_def __initdata am33xx_es1_0_opp_def_list[] = {
 	/* MPU OPP1 - OPP50 */
 	OPP_INITIALIZER("mpu", true,  275000000, AM33XX_VDD_MPU_OPP50_UV),
 	/* MPU OPP2 - OPP100 */
@@ -178,12 +179,88 @@ static struct omap_opp_def __initdata am
 	OPP_INITIALIZER("mpu", true,  720000000, AM33XX_VDD_MPU_OPPTURBO_UV),
 };
 
+#define AM33XX_ES2_0_VDD_MPU_OPP50_UV		950000
+#define AM33XX_ES2_0_VDD_MPU_OPP100_UV		1100000
+#define AM33XX_ES2_0_VDD_MPU_OPP120_UV		1200000
+#define AM33XX_ES2_0_VDD_MPU_OPPTURBO_UV	1260000
+#define AM33XX_ES2_0_VDD_MPU_OPPNITRO_UV	1325000
+
+static struct omap_opp_def __initdata am33xx_es2_0_opp_def_list[] = {
+	/* MPU OPP1 - OPP50 */
+	OPP_INITIALIZER("mpu", true,  300000000, AM33XX_ES2_0_VDD_MPU_OPP50_UV),
+	/* MPU OPP2 - OPP100 */
+	OPP_INITIALIZER("mpu", true,  600000000,
+				AM33XX_ES2_0_VDD_MPU_OPP100_UV),
+	/* MPU OPP3 - OPP120 */
+	OPP_INITIALIZER("mpu", true,  720000000,
+				AM33XX_ES2_0_VDD_MPU_OPP120_UV),
+	/* MPU OPP4 - OPPTurbo */
+	OPP_INITIALIZER("mpu", true, 800000000,
+				AM33XX_ES2_0_VDD_MPU_OPPTURBO_UV),
+	/* MPU OPP5 - OPPNitro */
+	OPP_INITIALIZER("mpu", false, 1000000000,
+				AM33XX_ES2_0_VDD_MPU_OPPNITRO_UV),
+};
+
+#define AM33XX_ES2_1_VDD_MPU_OPP50_UV		950000
+#define AM33XX_ES2_1_VDD_MPU_OPP100_UV		1100000
+#define AM33XX_ES2_1_VDD_MPU_OPP120_UV		1200000
+#define AM33XX_ES2_1_VDD_MPU_OPPTURBO_UV	1260000
+#define AM33XX_ES2_1_VDD_MPU_OPPNITRO_UV	1325000
+
+#define OPP_50_300_INDEX	0
+#define OPP_100_300_INDEX	1
+#define OPP_100_600_INDEX	2
+#define OPP_120_720_INDEX	3
+#define OPP_TURBO_800_INDEX	4
+#define OPP_NITRO_1GHZ_INDEX	5
+
+/* From AM335x TRM, SPRUH73H, Section 9.3.50 */
+#define AM33XX_EFUSE_SMA_OFFSET	0x7fc
+
+/*
+ * Bits [12:0] are OPP Disabled bits,
+ * 1 = OPP is disabled and not available,
+ * 0 = OPP available.
+ */
+#define MAX_FREQ_MASK		0x1fff
+#define MAX_FREQ_SHFT		0
+
+#define OPP_50_300MHZ_BIT		(0x1 << 4)
+#define OPP_100_300MHZ_BIT		(0x1 << 5)
+#define OPP_100_600MHZ_BIT		(0x1 << 6)
+#define OPP_120_720MHZ_BIT		(0x1 << 7)
+#define OPP_TURBO_800MHZ_BIT		(0x1 << 8)
+#define OPP_NITRO_1GHZ_BIT		(0x1 << 9)
+
+static struct omap_opp_def __initdata am33xx_es2_1_opp_list[] = {
+	/* MPU OPP1 - OPP50-300MHz */
+	OPP_INITIALIZER("mpu", false,  300000000,
+				AM33XX_ES2_1_VDD_MPU_OPP50_UV),
+	/* MPU OPP1 - OPP100-300MHz (used for ZCE) */
+	OPP_INITIALIZER("mpu", false,  300000000,
+				AM33XX_ES2_1_VDD_MPU_OPP100_UV),
+	/* MPU OPP2 - OPP100 */
+	OPP_INITIALIZER("mpu", false,  600000000,
+				AM33XX_ES2_1_VDD_MPU_OPP100_UV),
+	/* MPU OPP3 - OPP120 */
+	OPP_INITIALIZER("mpu", false,  720000000,
+				AM33XX_ES2_1_VDD_MPU_OPP120_UV),
+	/* MPU OPP3 - OPPTurbo */
+	OPP_INITIALIZER("mpu", false, 800000000,
+				AM33XX_ES2_1_VDD_MPU_OPPTURBO_UV),
+	/* MPU OPP4 - OPPNitro */
+	OPP_INITIALIZER("mpu", false, 1000000000,
+				AM33XX_ES2_1_VDD_MPU_OPPNITRO_UV),
+};
+
 /**
  * omap3_opp_init() - initialize omap3 opp table
  */
 int __init omap3_opp_init(void)
 {
 	int r = -ENODEV;
+	u32 rev, val, max_freq;
 
 	if (!cpu_is_omap34xx())
 		return r;
@@ -191,12 +268,77 @@ int __init omap3_opp_init(void)
 	if (cpu_is_omap3630())
 		r = omap_init_opp_table(omap36xx_opp_def_list,
 			ARRAY_SIZE(omap36xx_opp_def_list));
-	else if (cpu_is_am33xx())
-		r = omap_init_opp_table(am33xx_opp_def_list,
-			ARRAY_SIZE(am33xx_opp_def_list));
-	else
+	else if (cpu_is_am33xx()) {
+		rev = omap_rev();
+		switch (rev) {
+		case AM335X_REV_ES1_0:
+			r = omap_init_opp_table(am33xx_es1_0_opp_def_list,
+				ARRAY_SIZE(am33xx_es1_0_opp_def_list));
+			break;
+
+		case AM335X_REV_ES2_1:
+			/*
+			 * First read efuse sma reg to detect package type and
+			 * supported frequency
+			 */
+			val =
+			readl(AM33XX_CTRL_REGADDR(AM33XX_EFUSE_SMA_OFFSET));
+
+			if (!(val & MAX_FREQ_MASK)) {
+				/*
+				* if mpu max freq is not populated, fall back to
+				* PG 2.0 OPP settings.
+				*/
+				r =
+				omap_init_opp_table(am33xx_es2_0_opp_def_list,
+					ARRAY_SIZE(am33xx_es2_0_opp_def_list));
+				break;
+			}
+
+			/*
+			 * 1 = OPP is disabled and not available,
+			 * 0 = OPP available.
+			 */
+			max_freq = (~val & MAX_FREQ_MASK);
+
+			if (max_freq & OPP_50_300MHZ_BIT)
+				am33xx_es2_1_opp_list[OPP_50_300_INDEX].
+					default_available = true;
+
+			if (max_freq & OPP_100_300MHZ_BIT)
+				am33xx_es2_1_opp_list[OPP_100_300_INDEX].
+					default_available = true;
+
+			if (max_freq & OPP_100_600MHZ_BIT)
+				am33xx_es2_1_opp_list[OPP_100_600_INDEX].
+					default_available = true;
+
+			if (max_freq & OPP_120_720MHZ_BIT)
+				am33xx_es2_1_opp_list[OPP_120_720_INDEX].
+					default_available = true;
+
+			if (max_freq & OPP_TURBO_800MHZ_BIT)
+				am33xx_es2_1_opp_list[OPP_TURBO_800_INDEX].
+					default_available = true;
+
+			if (max_freq & OPP_NITRO_1GHZ_BIT)
+				am33xx_es2_1_opp_list[OPP_NITRO_1GHZ_INDEX].
+					default_available = true;
+
+			r = omap_init_opp_table(am33xx_es2_1_opp_list,
+				ARRAY_SIZE(am33xx_es2_1_opp_list));
+			break;
+
+		case AM335X_REV_ES2_0:
+		/* FALLTHROUGH */
+		default:
+			r = omap_init_opp_table(am33xx_es2_0_opp_def_list,
+				ARRAY_SIZE(am33xx_es2_0_opp_def_list));
+		}
+	} else {
 		r = omap_init_opp_table(omap34xx_opp_def_list,
 			ARRAY_SIZE(omap34xx_opp_def_list));
+	}
 
 	return r;
 }
diff -rupN orig/arch/arm/plat-omap/include/plat/cpu.h mod/arch/arm/plat-omap/include/plat/cpu.h
--- orig/arch/arm/plat-omap/include/plat/cpu.h	2014-04-15 19:03:00.000000000 +0200
+++ mod/arch/arm/plat-omap/include/plat/cpu.h	2014-05-03 20:27:42.000000000 +0200
@@ -437,6 +437,8 @@ IS_OMAP_TYPE(3517, 0x3517)
 
 #define AM335X_CLASS		0x33500034
 #define AM335X_REV_ES1_0	AM335X_CLASS
+#define AM335X_REV_ES2_0	(AM335X_CLASS | (0x1 << 8))
+#define AM335X_REV_ES2_1	(AM335X_CLASS | (0x2 << 8))
 
 #define OMAP443X_CLASS		0x44300044
 #define OMAP4430_REV_ES1_0	(OMAP443X_CLASS | (0x10 << 8))
