diff -rupN orig/drivers/input/touchscreen/ti_tscadc.c mod/drivers/input/touchscreen/ti_tscadc.c
--- orig/drivers/input/touchscreen/ti_tscadc.c	2014-04-06 15:15:00.000000000 +0200
+++ mod/drivers/input/touchscreen/ti_tscadc.c	2014-04-18 23:06:00.000000000 +0200
@@ -424,7 +424,6 @@ static irqreturn_t tsc_interrupt(int irq
 			z *= ts_dev->x_plate_resistance;
 			z /= z1;
 			z = (z + 2047) >> 12;
-
 			/*
 			 * Sample found inconsistent by debouncing
 			 * or pressure is beyond the maximum.
@@ -665,6 +664,7 @@ static	int __devinit tscadc_probe(struct
 		input_dev->keybit[BIT_WORD(BTN_TOUCH)] = BIT_MASK(BTN_TOUCH);
 		input_set_abs_params(input_dev, ABS_X, 0, MAX_12BIT, 0, 0);
 		input_set_abs_params(input_dev, ABS_Y, 0, MAX_12BIT, 0, 0);
+		input_set_abs_params(input_dev, ABS_PRESSURE, 0, MAX_12BIT, 0, 0);
 		/* register to the input system */
 		err = input_register_device(input_dev);
 		if (err)
