diff -rupN linux-orig/sound/soc/davinci/davinci-gisbc.c linux/sound/soc/davinci/davinci-gisbc.c
--- linux-orig/sound/soc/davinci/davinci-gisbc.c	1970-01-01 01:00:00.000000000 +0100
+++ linux/sound/soc/davinci/davinci-gisbc.c		2013-01-04 11:59:41.288550701 +0100
@@ -0,0 +1,121 @@
+/*
+ * ASoC driver for TI DAVINCI EVM platform
+ *
+ * Author:      Vladimir Barinov, <vbarinov@embeddedalley.com>
+ * Copyright:   (C) 2007 MontaVista Software, Inc., <source@mvista.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ */
+
+#include <linux/module.h>
+#include <linux/moduleparam.h>
+#include <linux/timer.h>
+#include <linux/interrupt.h>
+#include <linux/platform_device.h>
+#include <linux/i2c.h>
+#include <sound/core.h>
+#include <sound/pcm.h>
+#include <sound/soc.h>
+
+#include <asm/dma.h>
+#include <asm/mach-types.h>
+
+#include <asm/hardware/asp.h>
+#include <mach/edma.h>
+
+#include "davinci-pcm.h"
+#include "davinci-i2s.h"
+#include "davinci-mcasp.h"
+
+#define SYSCLK	24000000
+#define GISBC_AUDIO_FORMAT (SND_SOC_DAIFMT_DSP_B | \
+		SND_SOC_DAIFMT_CBM_CFM | SND_SOC_DAIFMT_IB_NF)
+
+static int gisbc_hw_params(struct snd_pcm_substream *substream,
+			 struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai = rtd->codec_dai;
+	struct snd_soc_dai *cpu_dai = rtd->cpu_dai;
+	int ret = 0;
+
+	/* set codec DAI configuration */
+	ret = snd_soc_dai_set_fmt(codec_dai, GISBC_AUDIO_FORMAT);
+	if (ret < 0)
+		return ret;
+
+	/* set cpu DAI configuration */
+	ret = snd_soc_dai_set_fmt(cpu_dai, GISBC_AUDIO_FORMAT);
+	if (ret < 0)
+		return ret;
+
+	/* set the codec system clock */
+	ret = snd_soc_dai_set_sysclk(codec_dai, 0, SYSCLK, SND_SOC_CLOCK_OUT);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static struct snd_soc_ops gisbc_ops = {
+	.hw_params = gisbc_hw_params,
+};
+
+static int gisbc_aic32x4_init(struct snd_soc_pcm_runtime *rtd)
+{
+	return 0;
+}
+
+static struct snd_soc_dai_link am335x_gisbc_dai = {
+	.name = "tlv320aic32x4",
+	.stream_name = "TLV320AIC32X4",
+	.cpu_dai_name = "davinci-mcasp.0",
+	.codec_dai_name = "tlv320aic32x4-hifi",
+	.codec_name = "tlv320aic32x4.1-0018",
+	.platform_name = "davinci-pcm-audio",
+	.init = gisbc_aic32x4_init,
+	.ops = &gisbc_ops,
+};
+
+static struct snd_soc_card am335x_snd_soc_card = {
+	.name = "AM335X GISBC",
+	.dai_link = &am335x_gisbc_dai,
+	.num_links = 1,
+};
+
+static struct platform_device *gisbc_snd_device;
+
+static int __init gisbc_init(void)
+{
+	int ret;
+
+	if (!machine_is_am335xevm())
+		return -EINVAL;
+
+	gisbc_snd_device = platform_device_alloc("soc-audio", 0);
+	if (!gisbc_snd_device) {
+		return -ENOMEM;
+	}
+
+	platform_set_drvdata(gisbc_snd_device, &am335x_snd_soc_card);
+	ret = platform_device_add(gisbc_snd_device);
+	if (ret) {
+		platform_device_put(gisbc_snd_device);
+	}
+
+	return ret;
+}
+
+static void __exit gisbc_exit(void)
+{
+	platform_device_unregister(gisbc_snd_device);
+}
+
+module_init(gisbc_init);
+module_exit(gisbc_exit);
+
+MODULE_AUTHOR("Marcin Niestroj");
+MODULE_DESCRIPTION("Grinn GISBC ASoC driver");
+MODULE_LICENSE("GPL");
diff -rupN linux-orig/sound/soc/davinci/davinci-mcasp.c linux/sound/soc/davinci/davinci-mcasp.c
--- linux-orig/sound/soc/davinci/davinci-mcasp.c	2012-11-09 16:16:08.000000000 +0100
+++ linux/sound/soc/davinci/davinci-mcasp.c	2013-01-04 11:59:41.288550701 +0100
@@ -494,6 +494,10 @@ static int davinci_mcasp_set_dai_fmt(str
 
 		mcasp_clr_bits(base + DAVINCI_MCASP_PDIR_REG,
 				ACLKX | AHCLKX | AFSX | ACLKR | AHCLKR | AFSR);
+
+		/* Set AHCLKX pin out - to provide MCLK for codec */
+		mcasp_set_bits(base + DAVINCI_MCASP_PDIR_REG,
+				AHCLKX);
 		break;
 
 	default:
diff -rupN linux-orig/sound/soc/davinci/Kconfig linux/sound/soc/davinci/Kconfig
--- linux-orig/sound/soc/davinci/Kconfig	2012-11-09 16:16:08.000000000 +0100
+++ linux/sound/soc/davinci/Kconfig		2013-01-04 11:59:41.280550699 +0100
@@ -41,6 +41,14 @@ config  SND_AM335X_SOC_EVM
 	help
 	  Say Y if you want to add support for SoC audio on
 	  AM335X EVM
+	  
+config SND_AM335X_SOC_GISBC
+	tristate "SoC Audio support for GISBC"
+	depends on SND_AM33XX_SOC
+	select SND_DAVINCI_SOC_MCASP
+	select SND_SOC_TLV320AIC32X4
+	help
+	  Say Y if you want to add support for SoC audio on GISBC
 
 choice
 	prompt "DM365 codec select"
diff -rupN linux-orig/sound/soc/davinci/Makefile linux/sound/soc/davinci/Makefile
--- linux-orig/sound/soc/davinci/Makefile	2012-11-09 16:16:08.000000000 +0100
+++ linux/sound/soc/davinci/Makefile		2013-01-04 11:59:41.288550701 +0100
@@ -12,6 +12,7 @@ obj-$(CONFIG_SND_DAVINCI_SOC_VCIF) += sn
 
 # DAVINCI Machine Support
 snd-soc-evm-objs := davinci-evm.o
+snd-soc-gisbc-objs := davinci-gisbc.o
 snd-soc-sffsdr-objs := davinci-sffsdr.o
 
 obj-$(CONFIG_SND_DAVINCI_SOC_EVM) += snd-soc-evm.o
@@ -19,4 +20,5 @@ obj-$(CONFIG_SND_DM6467_SOC_EVM) += snd-
 obj-$(CONFIG_SND_DA830_SOC_EVM) += snd-soc-evm.o
 obj-$(CONFIG_SND_DA850_SOC_EVM) += snd-soc-evm.o
 obj-$(CONFIG_SND_AM335X_SOC_EVM) += snd-soc-evm.o
+obj-$(CONFIG_SND_AM335X_SOC_GISBC) += snd-soc-gisbc.o
 obj-$(CONFIG_SND_DAVINCI_SOC_SFFSDR) += snd-soc-sffsdr.o
